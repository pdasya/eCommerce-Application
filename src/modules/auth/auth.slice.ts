import { createSlice } from '@reduxjs/toolkit';
import { RootState } from '@/store';

interface IAuthState {
  isAuthorized: boolean;
  id: number | null;
}

const initialState: IAuthState = {
  isAuthorized: false,
  id: null,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    authorize(state, action) {
      console.log('logged in');
      state.isAuthorized = true;
      state.id = action.payload.id;
    },
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    unauthorize(state, action) {
      console.log('logged out');
      state.isAuthorized = false;
      state.id = initialState.id;
    },
  },
});

export const { authorize, unauthorize } = authSlice.actions;
export const selectAuthorization = (state: RootState) => state.auth.isAuthorized;
export const authReducer = authSlice.reducer;
