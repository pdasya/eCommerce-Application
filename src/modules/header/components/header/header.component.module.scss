.toolbar {
  display: flex;
  width: 100%;
  box-sizing: border-box;
  justify-content: space-between;
  padding: 5px;
  gap: 10px;
}
