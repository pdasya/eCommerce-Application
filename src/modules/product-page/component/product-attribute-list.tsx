import React, { FC } from 'react';
import { Attribute } from '@commercetools/platform-sdk';

export const AttributeList: FC<Attribute> = (attribute: Attribute) => {
  const existAttribute = (item: string, field: string) => item.includes(field);
  const { name, value } = attribute;
  if (
    existAttribute(name, 'maker') ||
    existAttribute(name, 'volume') ||
    existAttribute(name, 'size')
  ) {
    return <p>{`${name}: ${value}`}</p>;
  }
  return '';
};
