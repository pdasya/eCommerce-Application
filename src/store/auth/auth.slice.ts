import { createSlice } from '@reduxjs/toolkit';
import { RootState } from '@/store';

interface IAuthState {
  isAuthorized: boolean;
  id: string;
  email: string;
}

const initialState: IAuthState = {
  isAuthorized: false,
  id: '',
  email: '',
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    authorize(state, action) {
      state.isAuthorized = true;
      state.id = action.payload.id;
      state.email = action.payload.email;
    },
    unauthorize(state, action) {
      state.isAuthorized = false;
      state.id = initialState.id;
      state.email = initialState.email;
    },
  },
});

export const { authorize, unauthorize } = authSlice.actions;
export const selectAuthorization = (state: RootState) => state.auth.isAuthorized;
export const authReducer = authSlice.reducer;
