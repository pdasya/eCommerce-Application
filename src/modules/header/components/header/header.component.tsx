import React, { FC } from 'react';
import { AppBar, CircularProgress, Toolbar } from '@mui/material';
import { selectAuthorization } from '@store/auth/auth.slice';
import { selectAuthPending } from '@store/misc/misc.slice';
import { NavBar } from '../nav-bar/nav-bar.component';
import { Logo } from '../logo/logo.component';
import { Cart } from '../cart/cart.component';
import styles from './header.component.module.scss';
import { SignInBar } from '../sign-in-bar/sign-in-bar.component';
import { UserBar } from '../user-bar/user-bar.component';
import { useAppSelector } from '@/hooks/use-app-selector.hook';

export const Header: FC = () => {
  const isAuthorized = useAppSelector(selectAuthorization);
  const isAuthPending = useAppSelector(selectAuthPending);

  return (
    <AppBar position="static">
      <Toolbar className={styles.toolbar}>
        <Logo />
        <NavBar />
        {isAuthPending ? (
          <CircularProgress color="warning" />
        ) : isAuthorized ? (
          <UserBar />
        ) : (
          <SignInBar />
        )}
        <Cart />
      </Toolbar>
    </AppBar>
  );
};
