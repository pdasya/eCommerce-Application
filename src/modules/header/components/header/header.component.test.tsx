import React from 'react';
import { act, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { authorize } from '@store/auth/auth.slice';
import { Header } from './header.component';
import { renderWithProviders } from '@/utils/render-with-providers.test-util';

describe('Header', () => {
  test('contains nav-links to all functional pages', () => {
    const navItemsNames = [
      'Home',
      'Catalog',
      'Product',
      'Cart',
      'Profile',
      'About',
      '404',
      'Error',
    ];

    renderWithProviders(<Header />);

    const navItems = screen.queryAllByRole('navigation');

    navItemsNames.forEach(name => {
      expect(navItems.find(item => within(item).queryByText(name))).toBeInTheDocument();
    });
  });

  test("render 'sing-in' / 'sign-up' options according to authorization status", () => {
    const { store } = renderWithProviders(<Header />);

    expect(screen.getByText('sign in')).toBeInTheDocument();
    expect(screen.getByText('register')).toBeInTheDocument();

    act(() => store.dispatch(authorize({})));

    expect(screen.queryByText('sign in')).not.toBeInTheDocument();
    expect(screen.queryByText('register')).not.toBeInTheDocument();
  });

  test('snapshot match', () => {
    const { asFragment } = renderWithProviders(<Header />);

    expect(asFragment()).toMatchSnapshot('header-snapshot');
  });
});
