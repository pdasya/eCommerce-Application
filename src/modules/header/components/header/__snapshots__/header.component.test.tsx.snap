// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot match: header-snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <a
        class=""
        href="/"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-elhp89-MuiSvgIcon-root"
          data-testid="RamenDiningIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 6H8V4.65l1-.12zm0 6H8V7h1zM6 7h1v5H6zm0-2.12 1-.12V6H6zM22 3V2L5 4v8H2c0 3.69 2.47 6.86 6 8.25V22h8v-1.75c3.53-1.39 6-4.56 6-8.25H10V7h12V6H10V4.41z"
          />
        </svg>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-m5x0z-MuiTypography-root"
        >
          FOOD STORE
        </h6>
      </a>
      <div>
        <div
          class="MuiBox-root css-1lvtzne"
        >
          <button
            aria-controls="menu-appbar"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-uwxyu2"
        >
          <a
            class=""
            href="/"
            role="navigation"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorWarning MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorWarning css-n98s6m-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Home
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            class=""
            href="/catalog/all"
            role="navigation"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-16dcpmj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Catalog
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            class=""
            href="/about"
            role="navigation"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-16dcpmj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              About Us
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
      </div>
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorWarning css-1rnp7nk-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
      <a
        class=""
        href="/cart"
      >
        <button
          aria-label="cart"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2M1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
    </div>
  </header>
</DocumentFragment>
`;
